Add an expando function fixer

h-joo opened the pull request.

h-joo committed:
Provide a fixer to add type for standalone expando functions

Update reference tests after the fixer fixes it and becomes identical
or update the diff reason

Signed-off-by: Hana Joo <hanajoo@google.com>

dragomirtitian commented on tests/baselines/reference/expandoFunctionNestedAssigments.js:
```diff
@@ -51,7 +51,8 @@ for(let f in (Foo.forIn = []) ){
         Foo.fromForInBodyNested = 1;
     }
-}
+}
+
```
Why so may white space differences?

h-joo replied:
I think it's caused by my editor configuration (I'm using vanilla vs code configuration) adding a newline at the end of the test file and thus it affects everything. Would you want me to roll back?

dragomirtitian replied:
I think it's best we roll back

h-joo replied:
I've rolled back, and noticed some things not being caught correctly so I needed to ad some changes in between. 
Please first review this PR :  https://github.com/bloomberg/TypeScript/pull/140/files and after that this one.

h-joo committed:
Update diff reasons for tests

Signed-off-by: Hana Joo <hanajoo@google.com>

dragomirtitian commented on src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:
```diff
@@ -234,16 +238,61 @@ function withChanges<T>(
+      if (!expandoFunc.name || elements.length === 0) return undefined;
+      const newProperties = []
+      for (const symbol of elements) {
+        if (!isIdentifierText(symbol.name, program.getCompilerOptions().target)) continue;
```
What is this case for ? keywords?

dragomirtitian commented on src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:
```diff
@@ -234,16 +238,61 @@ function withChanges<T>(
+      const newProperties = []
+      for (const symbol of elements) {
+        if (!isIdentifierText(symbol.name, program.getCompilerOptions().target)) continue;
+        newProperties.push(factory.createVariableStatement(
```
Shouldn't we check if the property is already defined in a non-expando way? For example if there already is a namespace and it has some of the members defined?

```ts

export function foo() {}
foo.x = 1;
foo.y = 1;

export namespace foo {
    export let x: string
}
```

dragomirtitian commented on src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:
```diff
@@ -234,16 +238,61 @@ function withChanges<T>(
+      for (const symbol of elements) {
+        if (!isIdentifierText(symbol.name, program.getCompilerOptions().target)) continue;
+        newProperties.push(factory.createVariableStatement(
+          /*modifiers*/ undefined,
```
Shouldn't the variable declaration be exported? If the variable is not exported from the namespace it will not be the same as the one that gets added through the assignment 

```ts

export function foo() {}
foo.x = 1;
foo.y = 1;

export namespace foo {
    let x: string// not the same x as the one outside
}
```

h-joo committed:
PR comment fixes

1. Add 'export' keyword to variables within the namespace.
2. Omit property creation in the namespace when it exists.
3. Update tests

Signed-off-by: Hana Joo <hanajoo@google.com>
